larq/optimizers_v2.py

```diff
@@ -5,6 +5,84 @@
 from copy import deepcopy
 
 
+@utils.register_keras_custom_object
+class OptimizerGroup(tf.keras.optimizers.Optimizer):
+    """Group of a full-precision and binary optimizer.
+
+    #TODO: Do we want to abstract this to `n` optimizers, not specifically one fp and
+    one bin? (I personally prefer the user-side simplicity of this current approach.)
+    """
+
+    def __init__(self, bin_optimizer, fp_optimizer, name="Group"):
+        super().__init__(name=name)  # TODO: Do we need to pass **kwargs?
+
+        type_err_msg = "Expected tf.keras.optimizers.Optimizer for `{}`, received `{}.`"
+        if not isinstance(bin_optimizer, tf.keras.optimizers.Optimizer):
+            raise TypeError(type_err_msg.format("bin_optimizer", type(bin_optimizer)))
+        if not isinstance(fp_optimizer, tf.keras.optimizers.Optimizer):
+            raise TypeError(type_err_msg.format("fp_optimizer", type(bin_optimizer)))
+
+        self.bin_optimizer = bin_optimizer
+        self.fp_optimizer = fp_optimizer
+
+    def apply_gradients(self, grads_and_vars, name=None):
+        bin_grads_and_vars, fp_grads_and_vars = [], []
+
+        for grad, var in grads_and_vars:
+            if self.is_binary(var):
+                bin_grads_and_vars.append((grad, var))
+            else:
+                fp_grads_and_vars.append((grad, var))
+
+        bin_train_op = self.bin_optimizer.apply_gradients(bin_grads_and_vars, name=name)
+        fp_train_op = self.fp_optimizer.apply_gradients(fp_grads_and_vars, name=name)
+
+        return tf.group(bin_train_op, fp_train_op, name="train_with_group")
+
+    def __getattr__(self, name):
+        if name == "lr":
+            return self.fp_optimizer.lr
+        return super().__getattr__(name)  # TODO is this still robust enough?
+
+    # TODO: Not sure if this needs to be here or forwarded to Bop.
+    def _get_decayed_hyper(self, name, var_dtype):
```

Comment:
Yes this should be a method of the individual optimizer. I don't think any external code tries to access this method but that may be something to check.

Reply:
Checked that this does not get called externally by TensorFlow or our code; removed from `BNNOptimizerDuo` in 3dd59a8.